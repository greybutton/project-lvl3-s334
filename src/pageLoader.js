import { promises as fsPromises } from 'fs';
import url from 'url';
import axios from 'axios';
import cheerio from 'cheerio';
import debug from 'debug';
import { keys, flatten } from 'lodash/fp';

import {
  makeDest, makeFilesDest, makeFullFilesDest, changeFileDest, makeFileDest,
} from './utils';

const log = debug('page-loader');

const tags = {
  link: {
    selector: 'link[href^="/"]:not([href^="//"], [href$=".xml"])',
    attr: 'href',
  },
  img: {
    selector: 'img[src^="/"]:not([src^="//"])',
    attr: 'src',
  },
  script: {
    selector: 'script[src^="/"]:not([src^="//"])',
    attr: 'src',
  },
};

const changeHtml = (data, dest) => {
  const $ = cheerio.load(data);

  keys(tags).forEach((key) => {
    $(tags[key].selector).map((i, elem) => {
      const fileDest = $(elem).attr(tags[key].attr);
      const newFileDest = changeFileDest(dest, fileDest);
      return $(elem).attr(tags[key].attr, newFileDest);
    });
  });

  const html = $.html();
  return html;
};

const getUrls = (data) => {
  const $ = cheerio.load(data);
  const links = keys(tags)
    .map(key => $(tags[key].selector).map((i, elem) => $(elem).attr(tags[key].attr)).get());
  const result = flatten(links);
  return result;
};

export default (link, options) => {
  const { protocol, hostname } = url.parse(link);
  if (!protocol && !hostname) {
    const message = `incorrent url ${link}`;
    throw new Error(message);
  }
  const dest = makeDest(link, options);
  const filesDest = makeFilesDest(link);
  const fullFilesDest = makeFullFilesDest(filesDest, options);

  let filesData = [];
  let html = '';

  return axios
    .get(link)
    .then(({ data }) => {
      html = changeHtml(data, filesDest);
      const fileUrls = getUrls(data).map(pathname => url.format({ protocol, hostname, pathname }));
      const filePromises = fileUrls.map(fileUrl => axios
        .get(fileUrl, { responseType: 'arraybuffer' })
        .then((response) => {
          log('load file', fileUrl);
          return response;
        }));
      return filePromises;
    })
    .then(promises => Promise.all(promises))
    .then((responses) => { filesData = responses; })
    .then(() => fsPromises.writeFile(dest, html))
    .then(() => log('create main file', dest))
    .then(() => fsPromises.mkdir(fullFilesDest))
    .then(() => log('create files directory', fullFilesDest))
    .then(() => filesData.map((response) => {
      const { data: fileData, config: { url: urlFile } } = response;
      const fileDest = makeFileDest(fullFilesDest, urlFile);
      fsPromises.writeFile(fileDest, fileData);
      return fileDest;
    }))
    .then(subFilesDests => subFilesDests.forEach(item => log('create sub file', item)));
};
