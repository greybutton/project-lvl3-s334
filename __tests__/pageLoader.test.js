import os from 'os';
import url from 'url';
import path from 'path';
import { promises as fsPromises } from 'fs';
import nock from 'nock';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';

import pageLoader from '../src';

const host = 'http://localhost';

axios.defaults.adapter = httpAdapter;

const makeDest = (link, options) => {
  const { hostname, pathname } = url.parse(link);
  const filename = hostname
    .concat(pathname)
    .replace(/\W+/g, '-')
    .concat('.html');
  const dest = path.join(options.output, filename);
  return dest;
};

describe('page loader', () => {
  test('test', async () => {
    const expected = 'test data';
    const testPath = '/test';
    const tempDir = await fsPromises.mkdtemp(
      path.resolve(os.tmpdir(), 'page-loader-greybutton-'),
    );
    const options = {
      output: tempDir,
    };
    const link = host.concat(testPath);
    const dest = makeDest(link, options);

    nock(host)
      .get(testPath)
      .reply(200, expected);

    await pageLoader(link, options);
    const received = await fsPromises.readFile(dest, { encoding: 'utf8' });

    expect(received).toBe(expected);
  });
  test('another test', async () => {
    const expected = 'another test data';
    const testPath = '/anothertest';
    const tempDir = await fsPromises.mkdtemp(
      path.resolve(os.tmpdir(), 'page-loader-greybutton-'),
    );
    const options = {
      output: tempDir,
    };
    const link = host.concat(testPath);
    const dest = makeDest(link, options);

    nock(host)
      .get(testPath)
      .reply(200, expected);

    await pageLoader(link, options);
    const received = await fsPromises.readFile(dest, { encoding: 'utf8' });

    expect(received).toBe(expected);
  });
});
