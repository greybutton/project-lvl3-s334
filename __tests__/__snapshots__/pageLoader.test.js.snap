// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error tests cant load resource 1`] = `"received is not a function"`;

exports[`error tests incorrect url 1`] = `"received is not a function"`;

exports[`error tests wrong output 1`] = `"received is not a function"`;
